File: chap5/sec5.1/MirankerLinigerS.m
function [t,y]=MirankerLinigerS(f,tspan,y0,N);
% MirankerLiniger1 solves ODEs using a predictor corrector method
%   [t,y]=MirankerLinigerS(f,tspan,y0,N) solves dy/dt=f(t,y) with
%   initial value y0 on the time interval tspan doing n steps of
%   a sequential predictor corrector method from Miranker and Liniger

dt=(tspan(2)-tspan(1))/N;
t=(tspan(1):dt:tspan(2))';   
y(:,1)=y0(:);                               % colon to make column vector
y(:,2)=y(:,1)+dt*f(t(1),y(:,1));            % start with an Euler step
for n=2:N-1,
  yp=y(:,n)+dt/2*(3*f(t(n),y(:,n))-f(t(n-1),y(:,n-1))); % sequential predictor
  y(:,n+1)=y(:,n)+dt/2*(f(t(n+1),yp)+f(t(n),y(:,n)));   % corrector method
end;

File: chap5/sec5.1/MirankerLinigerS.py
import numpy as np

def mirankerLinigerS(f, tspan, y0, N):
    """
    Solves ODEs using a predictor-corrector method.

    This function solves the differential equation dy/dt = f(t, y) with
    an initial value y0 over the time interval specified by tspan,
    using N steps of a sequential predictor-corrector method.

    Parameters
    ----------
    f : callable
        Right-hand side of the ODE, f(t, y).
    tspan : tuple
        Time interval (start, end).
    y0 : array
        Initial value.
    N : int
        Number of time steps.

    Returns
    -------
    t : array
        Time points.
    y : array
        Solution values at time points.
    """
    # Time step
    dt = (tspan[1] - tspan[0]) / N
    t = np.linspace(tspan[0], tspan[1], N+1)

    # Initialize solution array
    y = np.zeros((len(y0), N+1))
    y[:, 0] = y0

    # Start with an Euler step
    y[:, 1] = y[:, 0] + dt * f(t[0], y[:, 0])

    # Predictor-corrector method
    for n in range(1, N):
        # Predictor step
        yp = y[:, n] + dt / 2 * (3 * f(t[n], y[:, n]) - f(t[n-1], y[:, n-1]))

        # Corrector step
        y[:, n+1] = y[:, n] + dt / 2 * (f(t[n+1], yp) + f(t[n], y[:, n]))

    return t, y
